### PR TITLE
dbeaver/pro#1280 decrease consumers number if we lost producer(s) in …

### PR DESCRIPTION
**Pull request by LonwoLonwo**
…a task to avoid a non-editable task

2022-11-28 19_07_37-Error loading configuration

2022-11-28 19_07_43-Data Transfer -  export_case2

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/DataTransferSettings.java [hunk 1]
```diff
@@ -179,7 +180,17 @@ private void initializePipes(
             // Both producers and consumers specified
             // Processor belongs to non-database nodes anyway
             if (initProducers.length != initConsumers.length) {
-                throw new IllegalArgumentException("Producers number must match consumers number");
+                // Something went wrong
+                if (initProducers.length < initConsumers.length && initProducers[0] instanceof DatabaseTransferProducer) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/DataTransferSettings.java:184]: Is it possible, by any chance, to have an array of producers (`initProducers`) that has zero elements?

File: plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/DataTransferSettings.java [hunk 2]
```diff
@@ -179,7 +180,17 @@ private void initializePipes(
...
+                    // In this case, we had data transfer from table(s) to a file or another container.
+                    // But, probably, table(s) were deleted, and we lost our producer(s).
+                    // Usually, consumers do not have any special info, so we can delete extra items.
+                    IDataTransferConsumer<?,?>[] initConsumersCopy = new IDataTransferConsumer<?,?>[initProducers.length];
+                    System.arraycopy(initConsumers, 0, initConsumersCopy, 0, initProducers.length);
+                    initConsumers = initConsumersCopy;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/DataTransferSettings.java]: You can use the `Arrays#copyOf` here.

```suggestion
                    initConsumers = Arrays.copyOf(initConsumers, initProducers.length);
```

File: plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/DataTransferSettings.java (hunk 1 above)
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/DataTransferSettings.java]: Is it possible, by any chance, to have an array of producers (`initProducers`) that has zero elements?

Comment (LonwoLonwo) [reply plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/DataTransferSettings.java]: You can look in the code higher to know it.

**Review by ShadelessFox — APPROVED** [review]

**Comment by LonwoLonwo** [comment]
Cases:

- [x] Export to CSV task with one table. Delete table. Try to open "Edit task" on the task in the Task Manager.
- [x] Export to CSV task with two (or more) tables. Delete one table. Try to open "Edit task" on the task in the Task Manager. The Task Manager will open after the error message.
- [x] Export to CSV task with two (or more) tables. Delete one table. Try to run the task. The task will fail with the message.
- [x] Export to another schema task with two (or more) new tables. Delete one table. Try to open "Edit task" on the task in the Task Manager. The Task Manager will open after the error message.
- [x] Export to another schema task with two (or more) new tables. Delete one table. Try to run the task. The task will fail with the message.
- [x] Export to another schema task with two (or more) existing tables. Delete one table. Try to open "Edit task" on the task in the Task Manager. The Task Manager will open after the error message.
- [x] Export to another schema task with two (or more) existing tables. Delete one table. Try to run the task. The task will fail with the message.

**Comment by uslss** [comment]
verified

**Pull request merged**